chrome/content/zotero/itemTree.jsx

```diff
@@ -3810,6 +3811,27 @@ var ItemTree = class ItemTree extends LibraryTree {
 		}
 		return span;
 	}
+
+	async _refresh(){
+		// TODO: This is a hack to force a refresh of the item list
+		// when the columns change. It should be removed when the
+		// virtualized table is updated to handle this automatically.
+		this._columnsId = null;
+		const virtualizedTable = this.tree && this.tree._columns;
+		if (!virtualizedTable) {
+		  Zotero.debug("ItemTree is still loading. Refresh skipped.");
+		  return;
+		}
+		// Remove style list otherwise the change will not be updated
+		this.props.domEl.ownerDocument.querySelector(`.${virtualizedTable._styleKey}`)?.remove();
+		// Refresh to rebuild _columns
+		await this.refreshAndMaintainSelection();
+		// Construct a new virtualized-table, otherwise it will not be updated
+		this.tree._columns =
+		  new virtualizedTable.__proto__.constructor(this.tree);
+		// Refresh again to totally make the itemView updated
+		await this.refreshAndMaintainSelection();
```

Comment:
There is certainly a better way to do this. You should see what `ItemTree.changeCollectionTreeRow()` does when a `collectionTreeRow` with a different `visibilityGroup` is passed. E.g. you never need to recreate the VirtualizedTable, you likely need to call `this.forceUpdate`.

Moreover, there is a strict definition for what `refresh()` means in the context of the ItemTree. It retrieves all items appropriate for the view from the collectionTreeRow, and displays them. E.g. when a collectionTreeRow is changed.

I'm not entirely sure how much of the tree needs to be "refreshed" when adding a column, maybe it's just a matter of redrawing the tree with an additional column. You should investigate if setting `this._columnsId = null` and calling `this.forceUpdate()` is enough, and in that case, the function should be called `_resetColumns()`, `_redrawColumns()` or something like that. It also should be called from the `notify()` function as discussed in my main comment.

Reply:
Get it. I'll test this, thanks! The current refresh is not elegant and is overdone

Reply:
Done

Reply:
Should be better now